ostd/src/mm/page_table/cursor.rs

```diff
@@ -692,51 +691,47 @@ where
         assert!(end <= self.0.barrier_va.end);
 
         while self.0.va < end {
-            let cur_pte = self.0.read_cur_pte();
-            if !cur_pte.is_present() {
+            let cur_va = self.0.va;
+            let cur_level = self.0.level;
+            let cur_is_tracked = self.0.should_map_as_tracked();
+            let mut cur_entry = self.0.cur_entry();
+
+            // Skip if it is already absent.
+            if cur_entry.is_none() {
                 self.0.move_forward();
                 continue;
             }
 
-            // Go down if it's not a last node.
-            if !cur_pte.is_last(self.0.level) {
-                self.0.level_down();
-
-                // We have got down a level. If there's no mapped PTEs in
-                // the current node, we can go back and skip to save time.
-                if self.0.guards[(self.0.level - 1) as usize]
-                    .as_ref()
-                    .unwrap()
-                    .nr_children()
-                    == 0
-                {
-                    self.0.level_up();
+            // Go down if it's not a last entry.
+            if cur_entry.is_node() {
+                let Child::PageTable(pt) = cur_entry.to_owned() else {
+                    unreachable!("Already checked");
```

Comment:
This pattern seems to occur multiple times, so perhaps there can be a `cur_entry.to_page_table()` method (i.e., `is_node() && to_owned()`).

Reply:
There's only two cases of `is_node() && to_owned()` I think. Perhaps it's not worthwhile to add a method for it. To fix it we need either to add a method that internally uses `unreachable` to unwrap a `Child`, or uses unsafe code to check and clone the child. Either of them seemed not elegant.

Reply:
OK, I am not counting the number of `is_node() && to_owned()` exactly, just expressing a feeling while reviewing the code.

Since the count is only two, the current form is acceptable to me.